### ISSUE TITLE
Sporadic failed authentication when provider becomes null

### ISSUE THREAD

**Issue by lucaa** [xwikisas/integration-azure-oauth#16]
Setup: single wiki, azure authenticator version 1.5.4 installed with a single azure provider configured, on XWiki 14.4.x.

From time to time (not clear why or when), the XWiki login screen is displayed with a warning,  asking to login:
![image](https://user-images.githubusercontent.com/657716/193647325-2383214b-191a-4402-bda1-64a428bcda57.png)

This situation arrives all of a sudden in a working wiki (it doesn't happen following a restart or something).

When this happens, there is this in the logs:
```
2022-09-28 04:00:00,228 [DefaultQuartzScheduler_Worker-9] DEBUG i.i.IdentityOAuthEventListener - Event! org.xwiki.bridge.event.DocumentUpdatedEvent@ca7530a7 from XWiki.Notifications.NotificationEmailHourlySender
2022-09-28 04:00:00,228 [DefaultQuartzScheduler_Worker-9] DEBUG x.i.i.IdentityOAuthConfigTools - Obtained object null for  Object xwiki:IdentityOAuth.OAuthProviderClass^IdentityOAuth.OAuthProviderClass
2022-09-28 04:00:00,802 [Mail Prepare Thread] WARN  ultPeriodicMimeMessageIterator - No default email address is configured in the administration.
2022-09-28 04:00:01,726 [http-nio-8080-exec-5969 - https://<anonymized>/bin/login/XWiki/XWikiLogin?&identityOAuth=start&provider=AzureAD&browserLocation=https%3A%2F%2F<anonymized>%2Fbin%2Flogin%2FXWiki%2FXWikiLogin%3Bjsessionid%3D7434F34FF8B464054C92ACC920E9F223%3Fsrid%3Dxu2wphr3%26xredirect%3D%252Fbin%252Fget%252FMain%252FWebHome%253Fsrid%253Dxu2wphr3] DEBUG .i.DefaultIdentityOAuthManager - OAuthStart.
2022-09-28 04:00:01,727 [http-nio-8080-exec-5969 - https://<anonymized>/bin/login/XWiki/XWikiLogin?&identityOAuth=start&provider=AzureAD&browserLocation=https%3A%2F%2F<anonymized>%2Fbin%2Flogin%2FXWiki%2FXWikiLogin%3Bjsessionid%3D7434F34FF8B464054C92ACC920E9F223%3Fsrid%3Dxu2wphr3%26xredirect%3D%252Fbin%252Fget%252FMain%252FWebHome%253Fsrid%253Dxu2wphr3] WARN  .i.DefaultIdentityOAuthManager - Trouble at authorizing
com.xwiki.identityoauth.IdentityOAuthException: Provider "null" not found.
        at com.xwiki.identityoauth.internal.DefaultIdentityOAuthManager.getActiveProvider(DefaultIdentityOAuthManager.java:268)
        at com.xwiki.identityoauth.internal.DefaultIdentityOAuthManager.processOAuthStart(DefaultIdentityOAuthManager.java:295)
        at com.xwiki.identityoauth.IdentityOAuthScriptService.processOAuthStart(IdentityOAuthScriptService.java:61)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.doInvoke(UberspectImpl.java:571)
        at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.invoke(UberspectImpl.java:554)
        at org.apache.velocity.runtime.parser.node.ASTMethod.execute(ASTMethod.java:221)
        at org.apache.velocity.runtime.parser.node.ASTReference.execute(ASTReference.java:368)
        at org.apache.velocity.runtime.parser.node.ASTReference.value(ASTReference.java:704)
        at org.apache.velocity.runtime.parser.node.ASTExpression.value(ASTExpression.java:75)
        at org.apache.velocity.runtime.parser.node.ASTSetDirective.render(ASTSetDirective.java:242)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.parser.node.ASTElseIfStatement.render(ASTElseIfStatement.java:108)
        at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:190)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:439)
        at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:190)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:171)
        at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:439)
        at org.apache.velocity.Template.merge(Template.java:358)
        at org.apache.velocity.Template.merge(Template.java:262)
        at org.xwiki.velocity.internal.DefaultVelocityEngine.evaluate(DefaultVelocityEngine.java:280)
        at com.xpn.xwiki.render.DefaultVelocityManager.evaluate(DefaultVelocityManager.java:321)
        at org.xwiki.rendering.internal.macro.velocity.VelocityMacro.evaluateString(VelocityMacro.java:131)
        at org.xwiki.rendering.internal.macro.velocity.VelocityMacro.evaluateString(VelocityMacro.java:52)
        at org.xwiki.rendering.macro.script.AbstractScriptMacro.evaluateBlock(AbstractScriptMacro.java:278)
        at org.xwiki.rendering.macro.script.AbstractScriptMacro.execute(AbstractScriptMacro.java:181)
        at org.xwiki.rendering.macro.script.AbstractScriptMacro.execute(AbstractScriptMacro.java:57)
        at org.xwiki.rendering.internal.transformation.macro.MacroTransformation.transform(MacroTransformation.java:309)
        at org.xwiki.rendering.internal.transformation.DefaultRenderingContext.transformInContext(DefaultRenderingContext.java:183)
        at org.xwiki.rendering.async.internal.block.AbstractBlockAsyncRenderer.transform(AbstractBlockAsyncRenderer.java:76)
        at org.xwiki.rendering.async.internal.block.DefaultBlockAsyncRenderer.tranform(DefaultBlockAsyncRenderer.java:156)
        at org.xwiki.rendering.async.internal.block.DefaultBlockAsyncRenderer.execute(DefaultBlockAsyncRenderer.java:139)
        at org.xwiki.rendering.async.internal.block.AbstractBlockAsyncRenderer.render(AbstractBlockAsyncRenderer.java:157)
        at org.xwiki.uiextension.internal.WikiUIExtension.render(WikiUIExtension.java:200)
        at org.xwiki.rendering.async.internal.block.DefaultBlockAsyncRendererExecutor$DecoratorWrapper.render(DefaultBlockAsyncRendererExecutor.java:67)
        at org.xwiki.rendering.async.internal.block.DefaultBlockAsyncRendererExecutor$DecoratorWrapper.render(DefaultBlockAsyncRendererExecutor.java:49)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.lambda$syncRender$0(DefaultAsyncRendererExecutor.java:267)
        at com.xpn.xwiki.internal.security.authorization.DefaultAuthorExecutor.call(DefaultAuthorExecutor.java:98)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.syncRender(DefaultAsyncRendererExecutor.java:267)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.render(DefaultAsyncRendererExecutor.java:250)
        at org.xwiki.rendering.async.internal.block.DefaultBlockAsyncRendererExecutor.execute(DefaultBlockAsyncRendererExecutor.java:125)
        at org.xwiki.rendering.async.internal.block.DefaultBlockAsyncRendererExecutor.execute(DefaultBlockAsyncRendererExecutor.java:113)
        at org.xwiki.uiextension.internal.AbstractWikiUIExtension.execute(AbstractWikiUIExtension.java:120)
        at org.xwiki.uiextension.internal.AbstractWikiUIExtension.execute(AbstractWikiUIExtension.java:108)
        at jdk.internal.reflect.GeneratedMethodAccessor5749.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.doInvoke(UberspectImpl.java:571)
        at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.invoke(UberspectImpl.java:554)
        at org.apache.velocity.runtime.parser.node.ASTMethod.execute(ASTMethod.java:221)
        at org.apache.velocity.runtime.parser.node.ASTReference.execute(ASTReference.java:368)
        at org.apache.velocity.runtime.parser.node.ASTReference.value(ASTReference.java:704)
        at org.apache.velocity.runtime.parser.node.ASTExpression.value(ASTExpression.java:75)
        at org.apache.velocity.runtime.parser.node.ASTMethod.execute(ASTMethod.java:167)
        at org.apache.velocity.runtime.parser.node.ASTReference.execute(ASTReference.java:368)
        at org.apache.velocity.runtime.parser.node.ASTReference.render(ASTReference.java:492)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:171)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.directive.Foreach.renderBlock(Foreach.java:309)
        at org.apache.velocity.runtime.directive.Foreach.render(Foreach.java:279)
        at org.apache.velocity.runtime.parser.node.ASTDirective.render(ASTDirective.java:304)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.directive.VelocimacroProxy.render(VelocimacroProxy.java:218)
        at org.apache.velocity.runtime.directive.RuntimeMacro.render(RuntimeMacro.java:331)
        at org.apache.velocity.runtime.directive.RuntimeMacro.render(RuntimeMacro.java:261)
        at org.apache.velocity.runtime.parser.node.ASTDirective.render(ASTDirective.java:304)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:171)
        at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:439)
        at org.apache.velocity.Template.merge(Template.java:358)
        at org.apache.velocity.Template.merge(Template.java:262)
        at org.xwiki.velocity.internal.DefaultVelocityEngine.evaluate(DefaultVelocityEngine.java:280)
        at com.xpn.xwiki.render.DefaultVelocityManager.evaluate(DefaultVelocityManager.java:321)
        at com.xpn.xwiki.internal.template.VelocityTemplateEvaluator.evaluateContent(VelocityTemplateEvaluator.java:95)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.evaluateContent(TemplateAsyncRenderer.java:222)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.renderVelocity(TemplateAsyncRenderer.java:177)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.render(TemplateAsyncRenderer.java:138)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.render(TemplateAsyncRenderer.java:54)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.lambda$syncRender$0(DefaultAsyncRendererExecutor.java:267)
        at com.xpn.xwiki.internal.security.authorization.DefaultAuthorExecutor.call(DefaultAuthorExecutor.java:98)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.syncRender(DefaultAsyncRendererExecutor.java:267)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.render(DefaultAsyncRendererExecutor.java:250)
        at org.xwiki.rendering.async.internal.block.DefaultBlockAsyncRendererExecutor.render(DefaultBlockAsyncRendererExecutor.java:154)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.render(InternalTemplateManager.java:773)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.renderFromSkin(InternalTemplateManager.java:746)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.renderFromSkin(InternalTemplateManager.java:726)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.render(InternalTemplateManager.java:712)
        at com.xpn.xwiki.internal.template.DefaultTemplateManager.render(DefaultTemplateManager.java:91)
        at com.xpn.xwiki.internal.template.DefaultTemplateManager.render(DefaultTemplateManager.java:85)
        at com.xpn.xwiki.XWiki.evaluateTemplate(XWiki.java:2536)
        at com.xpn.xwiki.XWiki.parseTemplate(XWiki.java:2514)
        at com.xpn.xwiki.api.XWiki.parseTemplate(XWiki.java:997)
        at jdk.internal.reflect.GeneratedMethodAccessor5750.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.doInvoke(UberspectImpl.java:571)
        at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.invoke(UberspectImpl.java:554)
        at org.apache.velocity.runtime.parser.node.ASTMethod.execute(ASTMethod.java:221)
        at org.apache.velocity.runtime.parser.node.ASTReference.execute(ASTReference.java:368)
        at org.apache.velocity.runtime.parser.node.ASTReference.render(ASTReference.java:492)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.directive.VelocimacroProxy.render(VelocimacroProxy.java:218)
        at org.apache.velocity.runtime.directive.RuntimeMacro.render(RuntimeMacro.java:331)
        at org.apache.velocity.runtime.directive.RuntimeMacro.render(RuntimeMacro.java:261)
        at org.apache.velocity.runtime.parser.node.ASTDirective.render(ASTDirective.java:304)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:171)
        at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:439)
        at org.apache.velocity.Template.merge(Template.java:358)
        at org.apache.velocity.Template.merge(Template.java:262)
        at org.xwiki.velocity.internal.DefaultVelocityEngine.evaluate(DefaultVelocityEngine.java:280)
        at com.xpn.xwiki.render.DefaultVelocityManager.evaluate(DefaultVelocityManager.java:321)
        at com.xpn.xwiki.internal.template.VelocityTemplateEvaluator.evaluateContent(VelocityTemplateEvaluator.java:95)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.evaluateContent(TemplateAsyncRenderer.java:222)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.renderVelocity(TemplateAsyncRenderer.java:177)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.render(TemplateAsyncRenderer.java:138)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.render(TemplateAsyncRenderer.java:54)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.lambda$syncRender$0(DefaultAsyncRendererExecutor.java:267)
        at com.xpn.xwiki.internal.security.authorization.DefaultAuthorExecutor.call(DefaultAuthorExecutor.java:98)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.syncRender(DefaultAsyncRendererExecutor.java:267)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.render(DefaultAsyncRendererExecutor.java:250)
        at org.xwiki.rendering.async.internal.block.DefaultBlockAsyncRendererExecutor.render(DefaultBlockAsyncRendererExecutor.java:154)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.render(InternalTemplateManager.java:773)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.renderFromSkin(InternalTemplateManager.java:746)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.renderFromSkin(InternalTemplateManager.java:726)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.render(InternalTemplateManager.java:712)
        at com.xpn.xwiki.internal.template.DefaultTemplateManager.render(DefaultTemplateManager.java:91)
        at com.xpn.xwiki.internal.template.DefaultTemplateManager.render(DefaultTemplateManager.java:85)
        at com.xpn.xwiki.XWiki.evaluateTemplate(XWiki.java:2536)
        at com.xpn.xwiki.XWiki.parseTemplate(XWiki.java:2514)
        at com.xpn.xwiki.api.XWiki.parseTemplate(XWiki.java:997)
        at jdk.internal.reflect.GeneratedMethodAccessor5750.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.doInvoke(UberspectImpl.java:571)
        at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.invoke(UberspectImpl.java:554)
        at org.apache.velocity.runtime.parser.node.ASTMethod.execute(ASTMethod.java:221)
        at org.apache.velocity.runtime.parser.node.ASTReference.execute(ASTReference.java:368)
        at org.apache.velocity.runtime.parser.node.ASTReference.render(ASTReference.java:492)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.directive.VelocimacroProxy.render(VelocimacroProxy.java:218)
        at org.apache.velocity.runtime.directive.RuntimeMacro.render(RuntimeMacro.java:331)
        at org.apache.velocity.runtime.directive.RuntimeMacro.render(RuntimeMacro.java:261)
        at org.apache.velocity.runtime.parser.node.ASTDirective.render(ASTDirective.java:304)
        at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:147)
        at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:439)
        at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:190)
        at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:439)
        at org.apache.velocity.Template.merge(Template.java:358)
        at org.apache.velocity.Template.merge(Template.java:262)
        at org.xwiki.velocity.internal.DefaultVelocityEngine.evaluate(DefaultVelocityEngine.java:280)
        at com.xpn.xwiki.render.DefaultVelocityManager.evaluate(DefaultVelocityManager.java:321)
        at com.xpn.xwiki.internal.template.VelocityTemplateEvaluator.evaluateContent(VelocityTemplateEvaluator.java:95)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.evaluateContent(TemplateAsyncRenderer.java:222)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.renderVelocity(TemplateAsyncRenderer.java:177)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.render(TemplateAsyncRenderer.java:138)
        at com.xpn.xwiki.internal.template.TemplateAsyncRenderer.render(TemplateAsyncRenderer.java:54)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.lambda$syncRender$0(DefaultAsyncRendererExecutor.java:267)
        at com.xpn.xwiki.internal.security.authorization.DefaultAuthorExecutor.call(DefaultAuthorExecutor.java:98)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.syncRender(DefaultAsyncRendererExecutor.java:267)
        at org.xwiki.rendering.async.internal.DefaultAsyncRendererExecutor.render(DefaultAsyncRendererExecutor.java:250)
        at org.xwiki.rendering.async.internal.block.DefaultBlockAsyncRendererExecutor.render(DefaultBlockAsyncRendererExecutor.java:154)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.render(InternalTemplateManager.java:773)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.renderFromSkin(InternalTemplateManager.java:746)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.renderFromSkin(InternalTemplateManager.java:726)
        at com.xpn.xwiki.internal.template.InternalTemplateManager.render(InternalTemplateManager.java:712)
        at com.xpn.xwiki.internal.template.DefaultTemplateManager.render(DefaultTemplateManager.java:91)
        at com.xpn.xwiki.internal.template.DefaultTemplateManager.render(DefaultTemplateManager.java:85)
        at com.xpn.xwiki.XWiki.evaluateTemplate(XWiki.java:2536)
        at com.xpn.xwiki.web.Utils.parseTemplate(Utils.java:179)
        at com.xpn.xwiki.web.XWikiAction.execute(XWikiAction.java:601)
        at com.xpn.xwiki.web.XWikiAction.execute(XWikiAction.java:289)
        at com.xpn.xwiki.web.LegacyActionServlet.service(LegacyActionServlet.java:114)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at com.xpn.xwiki.web.ActionFilter.doFilter(ActionFilter.java:122)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.xwiki.wysiwyg.filter.ConversionFilter.doFilter(ConversionFilter.java:61)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.xwiki.container.servlet.filters.internal.SetHTTPHeaderFilter.doFilter(SetHTTPHeaderFilter.java:63)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.xwiki.container.servlet.filters.internal.SavedRequestRestorerFilter.doFilter(SavedRequestRestorerFilter.java:208)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.xwiki.container.servlet.filters.internal.SetCharacterEncodingFilter.doFilter(SetCharacterEncodingFilter.java:111)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.xwiki.resource.servlet.RoutingFilter.doFilter(RoutingFilter.java:132)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
        at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:747)
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:688)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:375)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1639)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

A restart fixes the issue and login with Azure works again.

It's not clear why the provider is null at that point. From what I see in the code the list of providers is reloaded on application ready event and upon update of the page holding configuration. There is indeed an edit event just before this log, but the logs for it show that it didn't endup in the providers reload code.

**Comment by lucaa** [comment]
note: a customization was done in the authentication (in LoginExtension.vm) in order to redirect to the login with the Azure provider selected whenever the login page is reached. This should not impact the fact that the provider _becomes null_ after working for a while.

**Comment by lucaa** [comment]
A grep in the logs showed no occurences for 'Reloading IdentityOAuth providers' , which is logged by the oauth backend when it reloads providers.

**Comment by mflorea** [comment]
The only reason for the **sudden** null provider for the "AzureAD" hint that I can see by reading the code is that [``rebuildProviders``](https://github.com/xwikisas/identity-oauth/blob/identity-oauth-1.4/api/src/main/java/com/xwiki/identityoauth/internal/DefaultIdentityOAuthManager.java#L198-L205) is called. It's the only place that modifies the map of providers. I see two cases:

* The problem doesn't persist on **page reload**: in this case the issue is that the map of provides is cleared before being refilled so if an authentication request arrives at the same time it can find the map of providers either empty or partially filled. Since ``DefaultIdentityOAuthManager`` is a singleton with a **state** it must be more careful with how its state is modified in a multi-threaded environment (multiple HTTP requests). The solution I see is to reduce the time between clear and refill, and to use ``synchronized``.
* The problem persists on page reload. In this case there could be two issues:
  * The map of provides is cleared but then there is an exception when refilling it, which leaves it either empty or in a partial state; the solution should be to clear the map of provides only after we have successfully called ``loadAndRebuildProviders()``.
  * The call to ``loadAndRebuildProviders()`` suddenly doesn't find the available providers. But this calls ``loadProviderConfigs`` so you should see either "Found providers: " or "Trouble at loading IdentityOAuthProvider-configurations." in the logs, somewhere before the problem occurs.

**Comment by lucaa** [comment]
In this case multiple users had the problem, so we can consider it persistant on page reload.

**Comment by lucaa** [comment]
We found the cause of this (with @mflorea ):
* it happens whenever a jar extension is uninstalled from extension manager
* the DefaultIdentityOAuthManager is reloaded at that time (because that's how extension manager works), so it loses its list of loaded providers
* the providers loading is done on ApplicationReadyEvent or DocumentUpdatedEvent of some specific docs, none of which is fired when this reloading from EM happens
* hence, after this operation the list of all providers is null and, of course, the AzureAD provider is also null.

This would be a "third case" of when providers are 'reloaded' (they're not actually reloaded).

The tricky part is that the debug logs of the Azure Authenticator _during the uninstallation_ are actually captured by the Extension Manager logs and displayed in the extension unistall logs. We had the following logs in the progress tab of unistallation of Polls app, which prove the reloading of the DefaultIdentityOAuthManager:

```
...
Successfully applied [UNINSTALL] for extension [XPoll Application (Pro) API 2.1.5] on namespace [] from previous extension(s) [com.xwiki.xpoll:application-xpoll-api/2.1.5]
Lifecycle to STOPPING
Lifecycle to STOPPED
IdentityOAuthManagerImpl initializing.
Lifecycle to INITIALIZED
Deleted document [Home » XPoll » XPollClass]
...
```

**Comment by lucaa** [comment]
The fix would be to add an additional event to the list of events on which reload of providers is happening, the proper event that is fired when this reloading of classes happens in the EM.

In the mean time the workaround is to make sure to restart the wiki when uninstalling jars.

**Comment by oanalavinia** [comment]
Fixed as part of #20 , which includes https://github.com/xwikisas/identity-oauth/issues/16
